File: cloud-collect-core/src/main/java/com/windea/demo/cloudcollect/core/service/impl/JwtUserDetailsService.java
package com.windea.demo.cloudcollect.core.service.impl;

import com.windea.demo.cloudcollect.core.domain.model.JwtUserDetails;
import com.windea.demo.cloudcollect.core.repository.UserRepository;
import org.springframework.security.core.userdetails.*;
import org.springframework.stereotype.Service;

/**
 * Jwt用户详情的服务。用于安全验证。
 */
@Service
public class JwtUserDetailsService implements UserDetailsService {
	private static final String message = "401 用户未找到！";

	private final UserRepository repository;

	public JwtUserDetailsService(UserRepository repository) {
		this.repository = repository;
	}


	@Override
	public UserDetails loadUserByUsername(String username) {
		var user = repository.findByUsername(username).orElseThrow(() -> new UsernameNotFoundException(message));
		return new JwtUserDetails(user);
	}
}

File: cloud-collect-core/src/main/kotlin/com/windea/demo/cloudcollect/core/service/impl/JwtUserDetailsService.kt
package com.windea.demo.cloudcollect.core.service.impl

import com.windea.demo.cloudcollect.core.domain.response.*
import com.windea.demo.cloudcollect.core.exception.*
import com.windea.demo.cloudcollect.core.repository.*
import org.springframework.security.core.userdetails.*
import org.springframework.stereotype.*

/**Jwt用户详情的服务。用于安全验证。*/
@Service
class JwtUserDetailsService(
	private val userRepository: UserRepository
) : UserDetailsService {
	override fun loadUserByUsername(username: String): UserDetails {
		val user = userRepository.findByUsername(username) ?: throw UserNotFoundException()
		return JwtUserDetails(user)
	}
	
	//登录方法向SecurityContext中存储的是validAuthentication，而非authentication
	//前者不包含密码，但包含权限信息，因此不需要实现UserDetailsPasswordService
}
